common/beerocks/bcl/source/beerocks_ucc_listener.cpp

```diff
@@ -277,23 +275,18 @@ bool beerocks_ucc_listener::create_cmdu(ieee1905_1::CmduMessageTx &cmdu_tx,
             return false;
         }
 
-        tlv_type = cmdu_tx.getNextTlvType();
-        while (tlv_type != int(ieee1905_1::eTlvType::TLV_END_OF_MESSAGE)) {
-            if (tlv_type == int(wfa_map::eTlvTypeMap::TLV_CHANNEL_PREFERENCE)) {
-                if (!cmdu_tx.addClass<wfa_map::tlvChannelPreference>()) {
-                    LOG(ERROR) << "addClass tlvChannelPreference has failed";
-                    err_string = err_internal;
-                    return false;
-                }
-            } else if (tlv_type == int(wfa_map::eTlvTypeMap::TLV_TRANSMIT_POWER_LIMIT)) {
-                if (!cmdu_tx.addClass<wfa_map::tlvTransmitPowerLimit>()) {
-                    LOG(ERROR) << "addClass tlvTransmitPowerLimit has failed";
-                    err_string = err_internal;
-                    return false;
-                }
-            }
-            tlv_type = cmdu_tx.getNextTlvType();
+        auto tlvChannelPreference = cmdu_tx.getClass<wfa_map::tlvChannelPreference>();
+        if (!tlvChannelPreference) {
+            LOG(ERROR) << "getClass<tlvChannelPreference> failed";
+            return false;
         }
+
+        auto tlvTransmitPowerLimit = cmdu_tx.getClass<wfa_map::tlvTransmitPowerLimit>();
+        if (!tlvTransmitPowerLimit) {
+            LOG(ERROR) << "getClass<tlvTransmitPowerLimit> failed";
+            return false;
```

Comment:
I don't think these checks are needed anymore. We only did it before to make sure the structures were parsed properly. Now the complete parsing is done as part of load(), so the addClass calls simply aren't needed. It's up to the functions that prepare the certification buffer to make sure that it contains the appropriate CMDUs.

So basically, I think that this entire function can be simplified to:

```
bool beerocks_ucc_listener::create_cmdu(ieee1905_1::CmduMessageTx &cmdu_tx,
                                        ieee1905_1::eMessageType message_type,
                                        std::string &err_string)
{
    switch (message_type) {
    case ieee1905_1::eMessageType::CHANNEL_SELECTION_REQUEST_MESSAGE:
    case ieee1905_1::eMessageType::COMBINED_INFRASTRUCTURE_METRICS_MESSAGE: {

        auto cmdu_header = cmdu_tx.load();
        if (!cmdu_header) {
            LOG(ERROR) << "load cmdu has failed";
            err_string = err_internal;
            return false;
        }
        // Force mid
        cmdu_header->message_id() = g_mid;
        
        return true;
    }

    default:
        // If message type handler is not implemented, then return false with empty error, so the
        // caller would know that it was not handled.
        err_string = std::string();
        return false;
    }
}
```

Reply:
Agree. I'll validate again the everything Ok on testbed before I'll merge it.

Reply:
I think we should use `getClass` here.
The reason for this is that we call `create_cmdu` method at line 790 in `beerocks_ucc_listener.cpp` in order to check if the current  CMDU message has a preprepared buffer which should be loaded, otherwise, we need to create it manually and use TLV data from the command if exists.
if we remove those checks we'll never know in which situation are we.

Reply:
I disagree. But since this whole certification buffer thing is a horrible hack anyway, let's not discuss it further.

Reply:
@CoralMalachi, Arnout has suggested a correct implementation IMO. We used `AddClass` before, to add the classes into the CMDU vector so it will be swapped on `finalize()`. Now that the data on the buffer is always swapped there is no need to do it anymore, so getting the classes is useless now. The only thing that needs to be done is to modify the `MID` on the CMDU header.

If I'm not mistaken, the data on the CMDU header we get is already swapped, so I think that we need to override its value with swapped value:
 `cmdu_header->message_id() = `swap_16(g_mid);`

Reply:
One important thing to note is that any `getClass` (except for cmdu header) will return `nullptr` after `cmdu_tx.load()` - so these checks are definitely not needed since they will always result with the same outcome.

Reply:
As @tomereli said since the `getClass` always returns `nullptr` we pass the test by accident. 
The `CHANNEL_SELECTION_REQUEST_MESSAGE` CMDU should be loaded from the certification buffer, but instead, it is created manually and use TLVs from the command (which do not exist).

To reveal this kind of bug, when a CMDU is not created from the certification buffer when it should be, update the test_flows to validate it in the last commit.